Remove awaiting Future and more Improvements

Faithcaio opened the pull request:
phew thats a big one..

This PR aims to
 - never await
 - never hide DB access (in templates / with assign operator / implicit conversions / json writes)
 - do fewer database queries - if the database could figure out something in one query it should do that
 - reduce None.get | Try.get etc. without checking
 - reduce crazy unreadable nesting

___

In more detail:
 - OreAction/OreRequest - always includes HeaderData (if not authed default with no permissions)
 - HeaderData contains User specific data (global permissions and infos for display in header)
 - JoinableData: OrganizationData and ProjectData (data for one orga/project
 - ScopedOrganizationData (Orga Permissions for current user)
 - ScopedProjectData (Project Permissions for current user and more userspecific data) 
 - UserData (User/OrgaUser permissions for currentuser and more data)
 - VersionData (ProjectData + more data)
 - replaced all xxx_= methods with setXxx (But they still do DB updates without returning Future)
 - removed all implicit conversions that query DB
 - removed (almost) all await
 - Wrote manual queries for:
   - Project its recommended Version and its Channel together (e.g. Home Page)
   - ReviewQueue (Version Author Project Channel) and Review and its user
   - Users for Orga Notifications (Orga Member Role User => User RoleType)
   - Bulk Insert Notifications (not everywhere)
   - MissingFiles (Project Version => Owner Name Version)
   - Project Pages (with parent)
   - QueryRole for Trust (ProjectMember ProjectRole)
   - Permission Logic (trustIn) (Project Members Orga OrgaMember Role => RoleType)
   - JsonWrites that made DB queries
     - Projects (Project Version Channel Tags)
     - Version (Verson Project Channel author Tags)
     - User (User Projects Stars)
 - other queries were only reordered and/or extracted to be in one place instead of spread out in templates etc. 

___

TODO: Review and Cleanup
 - [x] implicit Requests are not everywhere working because e.g. ProjectRequest is not OreRequest
 - [x] missing members in OreRestfulApi

I left out caching for a Future PR.
Also the queries can still be optimized in a lot of places. (looking at Future.sequence usages and other sequential selects)

___
nice to have maybe in a future PR?
 - caching of Data case classes 
 - and also invalidate when appropriate (maybe we need separate caches?)
 - actual case class models without variables - return futures on insert/update - single update for multiple columns
 - move queries out of controllers
 - bulk select everything when possible
 ...?

Faithcaio committed:
use the damn future

progwml6 commented:
@Faithcaio this PR has conflicts

phase commented:
I'd appreciate better commit messages since we're not squashing. I can fix the conflicts tonight if you're busy. There are definitely a lot of great improvements in here that will greatly improve the code quality and performance. :tada: :tada:

phase committed:
Merge branch 'master' into unawait

This includes the Play 2.6 updates.

Signed-off-by: Jadon Fowler <jadonflower@gmail.com>

phase commented:
I fixed the conflicts. :+1:

Katrix commented:
Maybe use `Future.successful` instead of `Future.apply` for pure values?

Faithcaio commented:
I know. Was a bit rusty with my scala in the beginning.
But its just a simple search and replace

Katrix commented on app/controllers/Application.scala:
```diff
@@ -116,30 +118,67 @@ final class Application @Inject()(data: DataHelper,
+      }
+    }
+    fullList map { list =>
+      val lists = list.partition(_._3 == 0)
```
Deconstruct tuple here so both lists can be named?

Katrix commented on app/controllers/Application.scala:
```diff
@@ -206,22 +246,32 @@ final class Application @Inject()(data: DataHelper,
+          case Some(id) =>
+            val reviews = this.service.access[Review](classOf[Review])
+              .filter(_.userId === id)
+              .map(_.take(20).map(review => { review ->
```
Curly braces for map

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -35,116 +41,176 @@ final class Reviews @Inject()(data: DataHelper,
+    val userTable = TableQuery[UserTable]
+
+    // Query Orga Members
+    val q1: Query[(Rep[Int], Rep[Option[RoleType]]), (Int, Option[RoleType]), scala.Seq]  = for {
```
Any reason for scala.Seq here?

pschichtel replied:
probably copied the infered type from intellij or similar. I'm wondering why the explicit type at all. Did it cause problems with the union down below?

Faithcaio replied:
yes IntelliJ was confused with the types so I helped it

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -35,116 +41,176 @@ final class Reviews @Inject()(data: DataHelper,
+    }
+
+    // Query version author
+    val q2: Query[(Rep[Int], Rep[Option[RoleType]]), (Int, Option[RoleType]), scala.Seq]  = for {
```
Any reason for scala.Seq here?

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -35,116 +41,176 @@ final class Reviews @Inject()(data: DataHelper,
-          this.service.insert(review)
-          Redirect(routes.Reviews.showReviews(author, slug, versionString))
+          val closeOldReview = version.mostRecentUnfinishedReview.flatMap {
+            case Some(oldreview) =>
```
Missing handling of None here.

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -35,116 +41,176 @@ final class Reviews @Inject()(data: DataHelper,
+    (Authenticated andThen PermissionAction[AuthRequest](ReviewProjects)).async { implicit request =>
+      withProjectAsync(author, slug) { implicit project =>
+          withVersionAsync(versionString) { version =>
+          version.reviewById(reviewId).map {
```
Wrong indentation

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -43,18 +43,16 @@ class Pages @Inject()(forms: OreForms,
-      val parentId = project.pages.find(equalsIgnoreCase(_.slug, parts(0))).map(_.id.getOrElse(-1)).getOrElse(-1)
-      val pages: Seq[Page] = project.pages.filter(equalsIgnoreCase(_.slug, parts(1))).seq
-      (pages.find(_.parentId == parentId), false)
+      project.pages.find(equalsIgnoreCase(_.slug, parts(0))).map(_.flatMap(_.id).getOrElse(-1)).flatMap(parentId => {
```
Curly braces for flatMap

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -66,11 +66,13 @@ class Versions @Inject()(stats: StatTracker,
-        Ok(views.view(project, version.channel, version))
+    withVersionAsync(versionString) { version =>
+      version.channel.map{ channel =>
+          this.stats.projectViewed { implicit request =>
```
Wrong indentation

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -34,23 +35,27 @@ trait Actions extends Calls with ActionHelpers {
-      Redirect(ShowHome)
+  def onUnauthorized(implicit request: Request[_], ec: ExecutionContext): Future[Result] = {
+    for {
+      noRedirect <- Future(request.flash.get("noRedirect"))
```
Any reason for putting a pure expression in the for comprehension?

Faithcaio committed:
fixing more awaits
introduces HeaderData/ProjectData/VersionData case classes for use in templates and caching

Katrix commented on app/db/impl/access/UserBase.scala:
```diff
@@ -121,14 +153,21 @@ class UserBase(override val service: ModelService,
-    getSession(token.value).map(_.user)
+  def current(implicit session: Request[_], ec: ExecutionContext): Future[Option[User]] = {
+    session.cookies.get("_oretoken") match {
+    case None => Future(None)
```
Wrong indentation

Katrix commented on app/models/user/User.scala:
```diff
@@ -331,10 +340,22 @@ case class User(override val id: Option[Int] = None,
+    this.service.getModelBase(classOf[UserBase]).current.flatMap {
+      case None => Future.successful(false)
+      case Some(user) =>
+        if ( user.equals(this)) Future.successful(true)
```
Remove space after parenthesis

Katrix commented on app/ore/project/ProjectMember.scala:
```diff
@@ -16,22 +18,31 @@ import ore.user.Member
+    * @return Top role
+    */
+  override def headRole(implicit ec: ExecutionContext): Future[ProjectRole] =
+    this.roles.map(role => {
```
Use curly brace for map

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -243,36 +255,47 @@ trait ProjectFactory {
 
-    Try(this.forums.await(this.forums.createProjectTopic(newProject)))
+    // Create the project and it's settings
+    for {
```
Why use two for comprehensions here?

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -177,33 +185,35 @@ class ProjectBase(override val service: ModelService,
 
     val channelCleanup = for {
-      _ <- rcUpdate
+      proj <- rcUpdate
```
Why use two for comprehensions here?

Katrix commented on app/ore/rest/OreRestfulApi.scala:
```diff
@@ -106,26 +104,68 @@ trait OreRestfulApi {
+      "tags"          ->  tags.map(toJson(_)),
+      "downloads"     ->  v.downloadCount
+    )
+    author.map(a => json.+(("author", JsString(a)))).getOrElse(json)
```
Any reason for not using infix `+` here? Also, fold is probably clearer here.

Katrix commented on app/util/PendingAction.scala:
```diff
@@ -10,11 +10,11 @@ trait PendingAction[R] {
     * Completes the action.
     */
-  def complete(): Try[R]
+  def complete(implicit ec: ExecutionContext): Future[R]
```
This is a side-effecting method, so reintroducing a the standard parenthesis would be nice.

Katrix commented on app/util/PendingAction.scala:
```diff
@@ -10,11 +10,11 @@ trait PendingAction[R] {
     * Cancels the action.
     */
-  def cancel()
+  def cancel(implicit ec: ExecutionContext)
```
This is a side-effecting method, so reintroducing a the standard parenthesis would be nice.

pschichtel replied:
I'd also remove the execution context from trait methods, I'd rather inject it into the implementations. Also: public methods should have explicit return types.

Faithcaio committed:
fix more errors
remove database access from templates

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -107,7 +109,7 @@ class Versions @Inject()(stats: StatTracker,
       implicit val project = request.project
       withVersion(versionString) { version =>
-        project.recommendedVersion = version
+        project.setRecommendedVersion( version)
```
Remove space after parenthesis here

Katrix commented on app/ore/rest/OreRestfulApi.scala:
```diff
@@ -49,26 +50,29 @@ trait OreRestfulApi {
-      } sortBy { case (p, v, c) =>
-        ordering
+      //categories.map(_.toSeq).map { cats =>
+      val filtered = cats.map { ca =>
```
A fold might be clearer here

Katrix commented on app/ore/rest/OreRestfulApi.scala:
```diff
@@ -266,7 +311,27 @@ trait OreRestfulApi {
+    userList.map(ul => toJson(ul.map(writeUser)))
+  }
+
+  def writeUser(user: User): JsObject = {
```
Is that signature correct, and if it is, why not use a Writes?

Katrix commented on app/util/DataHelper.scala:
```diff
@@ -34,14 +36,17 @@ final class DataHelper @Inject()(config: OreConfig,
     * Resets the application to factory defaults.
     */
-  def reset(): Unit = {
+  def reset(implicit ec: ExecutionContext): Unit = {
```
This method is not pure, and as such should have an extra set of parenthesis.

Faithcaio committed:
address some issues
remove more db access from templates

Faithcaio committed:
success in the future?

Faithcaio committed:
do not hide db updates

Faithcaio committed:
optimize user queries for REST

Faithcaio committed:
hide softdeleted projects in REST endpoints

Faithcaio committed:
reviewqueue query again
and fixing more compile errors

Faithcaio committed:
remove more db access from templates
and more fixing compile errors

Faithcaio committed:
fix more compile errors

Faithcaio committed:
adding Data for templates to requests

Faithcaio committed:
extracting more out of templates
it compiles btw

Faithcaio committed:
main page loads again

Katrix commented on app/controllers/Application.scala:
```diff
@@ -54,64 +81,54 @@ final class Application @Inject()(data: DataHelper,
+    val currentUserId = request.data.currentUser.flatMap(_.id).getOrElse(-1)
+
+    // TODO platform filter is not implemented
+    val pform = platform.flatMap(p => Platforms.values.find(_.name.equalsIgnoreCase(p)).map(_.asInstanceOf[Platform]))
```
Use collectFirst here instead of find and map for single traversal

Katrix commented on app/controllers/Application.scala:
```diff
@@ -54,64 +81,54 @@ final class Application @Inject()(data: DataHelper,
+
+    // TODO platform filter is not implemented
+    val pform = platform.flatMap(p => Platforms.values.find(_.name.equalsIgnoreCase(p)).map(_.asInstanceOf[Platform]))
+    val platformFilter = pform.map(actions.platformFilter).getOrElse(ModelFilter.Empty)
```
Fold?

Katrix commented on app/controllers/Application.scala:
```diff
@@ -54,64 +81,54 @@ final class Application @Inject()(data: DataHelper,
+    // TODO platform filter is not implemented
+    val pform = platform.flatMap(p => Platforms.values.find(_.name.equalsIgnoreCase(p)).map(_.asInstanceOf[Platform]))
+    val platformFilter = pform.map(actions.platformFilter).getOrElse(ModelFilter.Empty)
+    var categoryList: Seq[Category] = categories.map(Categories.fromString).map(_.toSeq).getOrElse(Seq.empty)
```
Fold?

Katrix commented on app/controllers/Application.scala:
```diff
@@ -54,64 +81,54 @@ final class Application @Inject()(data: DataHelper,
+    val pform = platform.flatMap(p => Platforms.values.find(_.name.equalsIgnoreCase(p)).map(_.asInstanceOf[Platform]))
+    val platformFilter = pform.map(actions.platformFilter).getOrElse(ModelFilter.Empty)
+    var categoryList: Seq[Category] = categories.map(Categories.fromString).map(_.toSeq).getOrElse(Seq.empty)
+    val q = query.map(queryString => s"%${queryString.toLowerCase}%").getOrElse("%")
```
Fold?

Katrix commented on app/controllers/Application.scala:
```diff
@@ -54,64 +81,54 @@ final class Application @Inject()(data: DataHelper,
+      (p.description.toLowerCase like q) ||
+      (p.ownerName.toLowerCase like q) ||
+      (p.pluginId.toLowerCase like q)
+    } drop offset take pageSize
```
Unsure if queries have it, but slice often does the job of both drop and take in one.

Faithcaio replied:
doesnt look like it does

Katrix commented on app/controllers/Application.scala:
```diff
@@ -120,56 +137,71 @@ final class Application @Inject()(data: DataHelper,
+      service.DB.db.run(queryReviews(list.map(_._1.id.get)).result).map { reviewList =>
+        reviewList.groupBy(_._1.versionId)
+      } map { reviewsByVersion =>
+        (list, reviewsByVersion)
```
Any specific reason for tuppling the result here with the list? It should still be available in the outer scope.

Katrix commented on app/controllers/Application.scala:
```diff
@@ -120,56 +137,71 @@ final class Application @Inject()(data: DataHelper,
+    } map { case (list, reviewsByVersion) =>
+      val reviewData = reviewsByVersion.mapValues { reviews =>
+
+        reviews.filter { case (review, _) =>
```
Maybe it would be better here to sort reviews first, and then find a specific one? It would also mean that you could use find instead of filter.

Katrix commented on app/controllers/Application.scala:
```diff
@@ -249,18 +306,18 @@ final class Application @Inject()(data: DataHelper,
             val reviews = this.service.access[Review](classOf[Review])
               .filter(_.userId === id)
-              .map(_.take(20).map(review => { review ->
+              .map(_.take(20).map { review => { review ->
```
Use a single curly brace here instead of one before and one after the function argument.

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -46,9 +49,11 @@ class Pages @Inject()(forms: OreForms,
     val parts = page.split("/")
     if (parts.size == 2) {
-      project.pages.find(equalsIgnoreCase(_.slug, parts(0))).map(_.flatMap(_.id).getOrElse(-1)).flatMap(parentId => {
+      project.pages.find(equalsIgnoreCase(_.slug, parts(0))).map {
```
collectFirst?

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -65,11 +70,14 @@ class Pages @Inject()(forms: OreForms,
-      case (None, b) => notFound
+    implicit val r = request.request
+    val rootPages: Seq[(Page, Seq[Page])] = null // TODO get pages
+  val pageCount: Int = 0  //TODO get pagecount
```
Indent

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -65,11 +70,14 @@ class Pages @Inject()(forms: OreForms,
+    val rootPages: Seq[(Page, Seq[Page])] = null // TODO get pages
+  val pageCount: Int = 0  //TODO get pagecount
+    withPage(project.p, page).flatMap {
+      case (Some(p), b) => this.stats.projectViewed(_ => Ok(views.view(project, rootPages, p, pageCount, b)))(request)
```
Any reason request is no longer marked as implicit, but instead passed in explicitly here?

Faithcaio replied:
that needs to change but its because I need a ProjectRequest and have an implicit OreRequest in scope

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -176,9 +191,12 @@ class Projects @Inject()(stats: StatTracker,
+    implicit val r = request.request
+    val rootPages = null // TODO root pages
+    val pageCount: Int = 0 // TODO page count
+    this.stats.projectViewed(request => Ok(views.pages.view(project, rootPages, project.p.homePage, pageCount)))(request)
```
Any reason request is no longer marked as implicit, but instead passed in explicitly here?

Faithcaio committed:
fix pages and perm check warning
full slug needs parent page

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -201,8 +219,9 @@ class Projects @Inject()(stats: StatTracker,
     */
-  def showDiscussion(author: String, slug: String) = ProjectAction(author, slug) { implicit request =>
-    this.stats.projectViewed(implicit request => Ok(views.discuss(request.project)))
+  def showDiscussion(author: String, slug: String) = ProjectAction(author, slug) async { request =>
```
Any reason request is no longer marked as implicit, but instead passed in explicitly here?

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -67,14 +70,16 @@ class Versions @Inject()(stats: StatTracker,
+        this.stats.projectViewed { request =>
+          val versionData: VersionData = null // TODO fill versionData
+          Ok(views.view(versionData))
+        }(request)
```
Any reason request is no longer marked as implicit, but instead passed in explicitly here?

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -183,98 +190,159 @@ trait Actions extends Calls with ActionHelpers {
+      implicit val r = request
+
+      p match {
+      case None => Future.successful(None)
```
Indent

Katrix commented on app/db/impl/access/UserBase.scala:
```diff
@@ -153,21 +155,20 @@ class UserBase(override val service: ModelService,
-    case Some(cookie) => getSession(cookie.value).flatMap {
-        case None => Future(None)
-        case Some(s) => s.user
+      case None => Future.successful(None)
```
flatMap

Faithcaio replied:
how? thats option future option

Katrix replied:
Ah, ok. In that case let's wait til OptionT, or you can use a fold (`session.cookies.get("_oretoken").fold[Future[Option[User]](Future.successful(None))(cookie => getSession(cookie.value).flatMap(_.user))`)

Katrix commented on app/models/admin/Review.scala:
```diff
@@ -137,3 +138,16 @@ case class Message(message: String, time: Long = System.currentTimeMillis(), act
 }
+
+
+object Review {
```
These orderings need better names

Katrix commented on app/models/admin/VisibilityChange.scala:
```diff
@@ -29,6 +31,13 @@ case class VisibilityChange(override val id: Option[Int] = None,
   def isResolved: Boolean = !resolvedAt.isEmpty
 
+  def created: Future[Option[User]] = {
+    if (createdBy.isEmpty) Future.successful(None)
```
I assume createdBy is an Option, in that case, fold.

Faithcaio committed:
fix orga and user pages not loading at all but still missing data

Katrix commented on app/models/viewhelper/HeaderData.scala:
```diff
@@ -1,27 +1,107 @@
+  private def getHeaderData(user: User) = {
+
+    Future.successful(
+    HeaderData(Some(user),
```
Indent

Katrix commented on app/models/viewhelper/OrganizationData.scala:
```diff
@@ -0,0 +1,30 @@
+                            permissions: Map[Permission, Boolean])
+  extends JoinableData[OrganizationRole, OrganizationMember, Organization] {
+
+  def o: Organization = joinable
```
Not a fan of the o name here personally

Katrix commented on app/models/viewhelper/ProjectData.scala:
```diff
@@ -12,30 +15,38 @@ import play.twirl.api.Html
+
+  def flagCount = flags.size
+
+  def p: Project = joinable
```
Not a fan of the p name here personally

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -105,10 +113,15 @@ class Pages @Inject()(forms: OreForms,
+        val pageCount = pages.size + pages.values.map(_.size).sum
+        val parentPage = if (pages.contains(p)) None
+        else {
+          pages.find(p => p._2.contains(p)).map(_._1)
```
collectFirst

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -194,9 +197,11 @@ class Projects @Inject()(stats: StatTracker,
+
+    projects.queryProjectPages(project.p) flatMap { pages =>
+      val pageCount = pages.size + pages.values.map(_.size).sum
+      this.stats.projectViewed(request => Ok(views.pages.view(project, pages, project.p.homePage, None, pageCount)))(request)
```
Any reason request is no longer marked as implicit, but instead passed in explicitly here?

Katrix commented on app/models/project/Page.scala:
```diff
@@ -135,7 +135,7 @@ case class Page(override val id: Option[Int] = None,
     * @return String
     */
-  def fullSlug: String = if (parentPage.isDefined) { s"${parentPage.get.slug}/${slug}" } else { slug }
+  def fullSlug(parentPage: Option[Page]): String = if (parentPage.isDefined) { s"${parentPage.get.slug}/$slug" } else { slug }
```
fold

Katrix commented on app/controllers/Application.scala:
```diff
@@ -55,91 +81,151 @@ final class Application @Inject()(data: DataHelper,
 
-      val pform = platform.flatMap(p => Platforms.values.find(_.name.equalsIgnoreCase(p)).map(_.asInstanceOf[Platform]))
-      val platformFilter = pform.map(actions.platformFilter).getOrElse(ModelFilter.Empty)
+      if (categoryList != null && Categories.visible.toSet.equals(categoryList.toSet))
```
Fairly certain that categoryList can't be null, only empty.

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -141,31 +156,37 @@ class Versions @Inject()(stats: StatTracker,
+          versions <- futureVersions
+          r <- this.stats.projectViewed { request =>
+            Ok(views.list(project, allChannels, versions, visibleNames, p))
+          }(request)
```
Any reason request is no longer marked as implicit, but instead passed in explicitly here?

Katrix commented on app/models/viewhelper/HeaderData.scala:
```diff
@@ -0,0 +1,112 @@
+
+  def of[A](request: Request[A])(implicit cache: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext,
+      service: ModelService): Future[HeaderData] = {
+    request.cookies.get("_oretoken") match {
```
flatMap

Faithcaio committed:
adress some issues

Faithcaio committed:
fill project data

Faithcaio committed:
provide more data

Faithcaio committed:
version data and pending project data

Faithcaio committed:
fix page sorting

Faithcaio committed:
more version data
icon route needs no additional project data

Faithcaio committed:
versions had no author in the past

Faithcaio committed:
separate Scoped Data from normal data
and add caching

Faithcaio committed:
fix channel list

Faithcaio committed:
remove await in version creation

Faithcaio committed:
remove more await
fix deploy key

Faithcaio committed:
visibility and logs

Faithcaio committed:
filling more missing data
invalidate cache (in some places)

felixoi commented on app/controllers/project/Projects.scala:
```diff
@@ -311,6 +315,7 @@ class Projects @Inject()(stats: StatTracker,
     * @return Project icon
     */
   def showIcon(author: String, slug: String) = Action async { implicit request =>
+    // TODO maybe instead of redirect cache this on ore?
```
I'd say it's a good idea.

Faithcaio committed:
invalidating cache is hard so we will do that later

Faithcaio committed:
Merge remote-tracking branch 'origin/master' into unawait

# Conflicts:
#	app/models/user/Organization.scala
#	app/ore/rest/OreRestfulApi.scala
#	app/ore/user/MembershipDossier.scala
#	app/views/projects/list.scala.html

Faithcaio committed:
fix after merge

Faithcaio committed:
fix hasUnreadNotifications

Katrix commented on app/controllers/project/Channels.scala:
```diff
@@ -96,29 +112,34 @@ class Channels @Inject()(forms: OreForms,
+          case None => Future.successful(NotFound)
+          case Some(channel) =>
+            for {
+              nonEmpty <- channel.versions.nonEmpty
```
These futures can be run in parallel

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -43,18 +47,18 @@ class Pages @Inject()(forms: OreForms,
-      } else {
-        (result, false)
+      project.pages.find((ModelFilter[Page](_.slug === parts(0)) +&& ModelFilter[Page](_.parentId === -1)).fn).flatMap {
+        case Some(r) => Future { (Some(r), false) }
```
Future.successfull

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -67,13 +71,20 @@ class Pages @Inject()(forms: OreForms,
+    implicit val r = request.request
+
+    withPage(data.project, page).flatMap {
+      case (None, b) => Future.successful(notFound)
```
If the b is not used here, mark it as ignored

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -67,13 +71,20 @@ class Pages @Inject()(forms: OreForms,
+      case (Some(p), b) =>
+        projects.queryProjectPages(data.project) flatMap { pages =>
+          val pageCount = pages.size + pages.map(_._2.size).sum
+          val parentPage = if (pages.contains(p)) None
```
Comparing different types.

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -67,13 +71,20 @@ class Pages @Inject()(forms: OreForms,
+          val pageCount = pages.size + pages.map(_._2.size).sum
+          val parentPage = if (pages.contains(p)) None
+          else pages.collectFirst { case (pp, page) if page.contains(p) => pp }
+          this.stats.projectViewed(_ => Ok(views.view(data, request.scoped, pages, p, parentPage, pageCount, b)))(cache, request)
```
request should be marked as implicit again here

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -83,21 +94,33 @@ class Pages @Inject()(forms: OreForms,
+    val p = parts.size match {
+      case 2 => data.project.pages.find(equalsIgnoreCase(_.slug, parts(0))).map(_.flatMap(_.id).getOrElse(-1))
+        .map((parts(1), _))
+      case _ => Future{(parts(0), -1)}
```
Future.successfull

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -83,21 +94,33 @@ class Pages @Inject()(forms: OreForms,
+    implicit val r = request.request
+    val data = request.data
+    val parts = pageName.split("/")
+    val p = parts.size match {
```
Unless we need to handle other sizes in the future, I think a normal if else would fit better here

Katrix commented on app/controllers/project/Pages.scala:
```diff
@@ -155,13 +186,16 @@ class Pages @Inject()(forms: OreForms,
-      this.service.access[Page](classOf[Page]).remove(optionPage._1.get)
+  def delete(author: String, slug: String, page: String) = PageEditAction(author, slug).async { implicit request =>
+    val data = request.data
+    for {
```
A normal map would probably read better here IMO

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -64,8 +58,15 @@ class Projects @Inject()(stats: StatTracker,
+      orgas <- request.user.organizations.all
+      createOrga <- Future.sequence(orgas.map(orga => request.user can CreateProject in orga))
+    } yield {
+      val createdOrgas = orgas zip createOrga filter (_._2) map (_._1)
```
use collect instead of filter and map

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -85,9 +86,9 @@ class Projects @Inject()(stats: StatTracker,
-              val project = this.factory.startProject(plugin)
-              project.cache()
-              val model = project.underlying
+              val pendingProject = this.factory.startProject(plugin)
```
Any reason for the rename here?

Faithcaio replied:
its a PendingProject it was confusing

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -104,12 +105,19 @@ class Projects @Inject()(stats: StatTracker,
       case Some(pending) =>
-        Ok(views.create(Some(pending)))
+        for {
+          orgas <- request.user.organizations.all
```
orgas and owner can be run in paralell.

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -120,27 +128,50 @@ class Projects @Inject()(stats: StatTracker,
+
+                val authors = pendingProject.file.meta.get.getAuthors.asScala
+                for {
+                  users <- Future.sequence(authors.filterNot(_.equals(currentUser.username)).map(this.users.withName))
```
All of these can be run in parallel

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -170,9 +201,14 @@ class Projects @Inject()(stats: StatTracker,
+
+    projects.queryProjectPages(data.project) flatMap { pages =>
+      val pageCount = pages.size + pages.map(_._2.size).sum
+      this.stats.projectViewed(request => Ok(views.pages.view(data, request.scoped, pages, data.project.homePage, None, pageCount)))(cache, request)
```
request should be implicit here.

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -382,8 +422,12 @@ class Projects @Inject()(stats: StatTracker,
-    Ok(views.settings(request.project))
+  def showSettings(author: String, slug: String) = SettingsEditAction(author, slug) async { request =>
+    implicit val r = request.request
+    val projectData: ProjectData = request.data
```
Any reason for the type annotation here?

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -463,18 +507,19 @@ class Projects @Inject()(stats: StatTracker,
+        hasErrors =>
+          Future.successful(FormError(self.showSettings(author, slug), hasErrors)),
+        formData => {
+          data.settings.save(data.project, formData).map { r =>
```
Mark r as ignored

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -540,17 +589,25 @@ class Projects @Inject()(stats: StatTracker,
+      implicit val r = request.request
+      val project = request.data.project
+      for {
+        changes <- project.visibilityChangesByDate
```
changedBy and logs can be parallelized

Katrix commented on app/controllers/project/Projects.scala:
```diff
@@ -578,11 +635,12 @@ class Projects @Inject()(stats: StatTracker,
     val comment = this.forms.NeedsChanges.bindFromRequest.get.trim
-    project.setVisibility(VisibilityTypes.SoftDelete, comment, request.user)
-    Redirect(ShowHome).withSuccess(this.messagesApi("project.deleted", project.name))
+    data.project.setVisibility(VisibilityTypes.SoftDelete, comment, request.user.id.get).map { vc =>
```
Mark vc as ignored

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -66,11 +70,18 @@ class Versions @Inject()(stats: StatTracker,
+        data <- VersionData.of(request, version)
+        response <- this.stats.projectViewed { request =>
+          Ok(views.view(data, request.scoped))
+        }(cache, request)
```
request should be marked as implicit here

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -141,31 +158,37 @@ class Versions @Inject()(stats: StatTracker,
+          versions <- futureVersions
+          r <- this.stats.projectViewed { request =>
+            Ok(views.list(data, request.scoped, allChannels, versions, visibleNames, p))
+          }(cache, request)
```
request should be marked as implicit here

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -177,9 +200,14 @@ class Versions @Inject()(stats: StatTracker,
+  def showCreator(author: String, slug: String) = VersionEditAction(author, slug).async { request =>
+    val data = request.data
+    implicit val r = request.request
+    for {
```
A normal map would probably be more clear here

Katrix commented on app/controllers/project/Versions.scala:
```diff
@@ -381,55 +420,61 @@ class Versions @Inject()(stats: StatTracker,
             warn.remove()
             false
           } else
             true
```
Move that true on to the previous line

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -34,23 +41,27 @@ trait Actions extends Calls with ActionHelpers {
-      Redirect(ShowHome)
+  def onUnauthorized(implicit request: Request[_], ec: ExecutionContext): Future[Result] = {
+    val noRedirect = request.flash.get("noRedirect")
+    for {
```
A normal map is probably more clear here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+
+  }
+
+  private def maybeAuthRequest[A](request: Request[A], futUser: Future[Option[User]])(implicit ec: ExecutionContext, asyncCacheApi: AsyncCacheApi,db: JdbcBackend#DatabaseDef): Future[Either[Result, AuthRequest[A]]] = {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
   }
 
-  def projectAction(author: String, slug: String)(implicit ec: ExecutionContext) = new ActionRefiner[Request, ProjectRequest] {
+  def projectAction(author: String, slug: String)(implicit  modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = new ActionRefiner[OreRequest, ProjectRequest] {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
-      .flatMap(processProject(_, this.users.current(request)))
-      .map(new ProjectRequest[A](_, request))
-      .toRight(notFound()(request))
+  private def maybeProjectRequest[A](r: OreRequest[A], project: Future[Option[Project]])(implicit  modelService: ModelService, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext): Future[Either[Result, ProjectRequest[A]]] = {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
-      Some(project)
-    else
-      None
+  private def toProjectRequest[T](project: Project)(f: (ProjectData, ScopedProjectData) => T)(implicit request: OreRequest[_], modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
-      None
+  private def toProjectRequest[T](project: Project)(f: (ProjectData, ScopedProjectData) => T)(implicit request: OreRequest[_], modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = {
+    for {
+      data <- ProjectData.of(project)
```
These two can be parallelized

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+    } else {
+      if (user.isDefined) {
+        for {
+          check1 <- canEditAndNeedChangeOrApproval(project, user)
```
Only one of these need to be completed. Check out `Future.firstCompletedOf`

Faithcaio replied:
Only one of them with true as a result. 
If the first is false I still need to wait for the other

Katrix replied:
Still possible, you just need to add a bit more work first.
```scala
val check1 = ???
val check2 = ???
Future.firstCompletedOf(Seq(check1.map(Left.apply), check2.map(Right.apply))).flatMap {
  case Left(false) => check2
  case Right(false) => check1
  case _ => Future.successful(true)
}
```

You can also probably save it for the EitherT and parallel stuff. The parallel stuff is built around the idea of the code above.

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+    }
+  }
+
+  def authedProjectActionImpl(project: Future[Option[Project]])(implicit  modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = new ActionRefiner[AuthRequest, AuthedProjectRequest] {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
   }
 
-  def authedProjectAction(author: String, slug: String)(implicit ec: ExecutionContext) = authedProjectActionImpl(this.projects.withSlug(author, slug))
+  def authedProjectAction(author: String, slug: String)(implicit  modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = authedProjectActionImpl(projects.withSlug(author, slug))
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+  def authedProjectAction(author: String, slug: String)(implicit  modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = authedProjectActionImpl(projects.withSlug(author, slug))
 
-  def authedProjectActionById(pluginId: String)(implicit ec: ExecutionContext) = authedProjectActionImpl(this.projects.withPluginId(pluginId))
+  def authedProjectActionById(pluginId: String)(implicit  modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = authedProjectActionImpl(projects.withPluginId(pluginId))
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+  def authedProjectActionById(pluginId: String)(implicit  modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = authedProjectActionImpl(projects.withPluginId(pluginId))
 
-  def organizationAction(organization: String)(implicit ec: ExecutionContext) = new ActionRefiner[Request, OrganizationRequest] {
+  def organizationAction(organization: String)(implicit modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = new ActionRefiner[OreRequest, OrganizationRequest] {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
   }
 
-  def authedOrganizationAction(organization: String)(implicit ec: ExecutionContext) = new ActionRefiner[AuthRequest, AuthedOrganizationRequest] {
+  def authedOrganizationAction(organization: String)(implicit modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = new ActionRefiner[AuthRequest, AuthedOrganizationRequest] {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+
+  }
+
+  private def maybeOrgaRequest[T](maybeOrga: Option[Organization])(f: (OrganizationData, ScopedOrganizationData) => T)(implicit request: OreRequest[_], modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = {
```
Newline somewhere in here

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+
+  private def maybeOrgaRequest[T](maybeOrga: Option[Organization])(f: (OrganizationData, ScopedOrganizationData) => T)(implicit request: OreRequest[_], modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef) = {
+    maybeOrga match {
+      case None =>Future.successful(Left(notFound()))
```
Space after arrow

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+      case None =>Future.successful(Left(notFound()))
+      case Some(orga) =>
+        for {
+          data <- OrganizationData.of(orga)
```
These two can be parallelized

Katrix commented on app/controllers/sugar/Actions.scala:
```diff
@@ -115,148 +134,233 @@ trait Actions extends Calls with ActionHelpers {
+    }
+  }
+
+  def getOrga(request: OreRequest[_], organization: String)(implicit modelService: ModelService, ec: ExecutionContext, asyncCacheApi: AsyncCacheApi, db: JdbcBackend#DatabaseDef): Future[Option[Organization]] = {
```
Newline somewhere in here

Katrix commented on app/controllers/ApiController.scala:
```diff
@@ -53,10 +59,12 @@ final class ApiController @Inject()(api: OreRestfulApi,
-      case "v1" => Ok(this.api.getProjectList(categories, sort, q, limit, offset))
-      case _ => NotFound
+      case "v1" => this.api.getProjectList(categories, sort, q, limit, offset).map(Ok(_))
+      case _ => Future {
```
Future.sucessfull

Katrix commented on app/controllers/ApiController.scala:
```diff
@@ -67,30 +75,33 @@ final class ApiController @Inject()(api: OreRestfulApi,
-      case "v1" => ApiResult(this.api.getProject(pluginId))
-      case _ => NotFound
+      case "v1" => this.api.getProject(pluginId).map(ApiResult)
+      case _ => Future {NotFound}
```
Future.sucessfull

Katrix commented on app/controllers/ApiController.scala:
```diff
@@ -138,65 +149,93 @@ final class ApiController @Inject()(api: OreRestfulApi,
+
+            val apiKeyExists: Future[Boolean] = this.service.DB.db.run(compiled(Deployment, formData.apiKey, projectData.project.id.get).result)
+            val dep = for {
+              apiKey <- apiKeyExists
```
These can be parallelized

Katrix commented on app/controllers/ApiController.scala:
```diff
@@ -138,65 +149,93 @@ final class ApiController @Inject()(api: OreRestfulApi,
+                    case Right(pendingVersion) =>
+                      pendingVersion.createForumPost = formData.createForumPost
+                      pendingVersion.channelName = formData.channel.name
+                      pendingVersion.complete.map { newVersion =>
```
This method has side effects, use parenthesis here

Katrix commented on app/controllers/Application.scala:
```diff
@@ -55,91 +82,167 @@ final class Application @Inject()(data: DataHelper,
+    val canHideProjects = request.data.globalPerm(HideProjects)
+    val currentUserId = request.data.currentUser.flatMap(_.id).getOrElse(-1)
+
+    // TODO ordering is not implemented
```
Are these 2 TODOs resolved?

Faithcaio replied:
Platform/Tag filter was not implemented before and wont be with this PR.

Katrix commented on app/controllers/Application.scala:
```diff
@@ -55,91 +82,167 @@ final class Application @Inject()(data: DataHelper,
+      users <- Future.sequence(flags.map(_.user))
+      projects <- Future.sequence(flags.map(_.project))
+      perms <- Future.sequence(projects.map { project =>
+        val perms =VisibilityTypes.values.map(_.permission).map { perm =>
```
Space after =

Katrix commented on app/controllers/Application.scala:
```diff
@@ -55,91 +82,167 @@ final class Application @Inject()(data: DataHelper,
+  def showFlags() = FlagAction.async { implicit request =>
+    for {
+      flags <- this.service.access[Flag](classOf[Flag]).filterNot(_.isResolved)
+      users <- Future.sequence(flags.map(_.user))
```
These can probably be paralellized

Faithcaio replied:
before that make it less queries. 1:1 relationship flag-user-project can be queried in a single query instead of 1+2n in this case.
same for a lot of cases where I used Future.sequence

Katrix commented on app/controllers/Application.scala:
```diff
@@ -149,18 +252,30 @@ final class Application @Inject()(data: DataHelper,
+  def showHealth() = (Authenticated andThen PermissionAction[AuthRequest](ViewHealth)) async { implicit request =>
+
+    for {
+      noTopicProjects <- projects.filter(p => p.topicId === -1 || p.postId === -1)
```
These can probably be parallelized

Katrix commented on app/controllers/Application.scala:
```diff
@@ -176,7 +291,8 @@ final class Application @Inject()(data: DataHelper,
   def reset() = (Authenticated andThen PermissionAction[AuthRequest](ResetOre)) { implicit request =>
     this.config.checkDebug()
-    this.data.reset()
+    this.data.reset
```
reset has side effects, add the parenthesis back

Katrix commented on app/controllers/Application.scala:
```diff
@@ -206,22 +323,32 @@ final class Application @Inject()(data: DataHelper,
+      case None => Future.successful(NotFound)
+      case Some(u) =>
+        val activities: Future[Seq[(Object, Option[Project])]] = u.id match {
+          case None => Future{Seq.empty}
```
Future.sucessfull

Katrix commented on app/controllers/Application.scala:
```diff
@@ -239,134 +366,197 @@ final class Application @Inject()(data: DataHelper,
-    val flagsClosed: Seq[(String, String)] = last10DaysCountQuery("project_flags", "resolved_at")
-
-    Ok(views.users.admin.stats(reviews, uploads, totalDownloads, unsafeDownloads, flagsOpen, flagsClosed))
+    for {
```
These can be parallelized

Katrix commented on app/controllers/Application.scala:
```diff
@@ -239,134 +366,197 @@ final class Application @Inject()(data: DataHelper,
+      case None => Future.successful(notFound)
+      case Some(u) =>
+        for {
+          userData <- getUserData(request, user)
```
Some of these can be parallelized

Katrix commented on app/controllers/Application.scala:
```diff
@@ -239,134 +366,197 @@ final class Application @Inject()(data: DataHelper,
+  def updateUser(userName: String) = UserAdminAction.async { implicit request =>
+    this.users.withName(userName).flatMap {
+      case None => Future.successful(NotFound)
+      case Some(user) => {
```
The curly brace here is not needed

Faithcaio committed:
add members to restapi projects again

Katrix commented on app/controllers/Application.scala:
```diff
@@ -239,134 +366,197 @@ final class Application @Inject()(data: DataHelper,
-    val futureApproval = projectSchema.collect(ModelFilter[Project](_.visibility === VisibilityTypes.NeedsApproval).fn, ProjectSortingStrategies.Default, -1, 0)
-    val projectApprovals = this.service.await(futureApproval).get
+    for {
+      projectApprovals <- projectSchema.collect(ModelFilter[Project](_.visibility === VisibilityTypes.NeedsApproval).fn, ProjectSortingStrategies.Default, -1, 0)
```
Some of the can be parallelized

Faithcaio committed:
Merge remote-tracking branch 'origin/master' into unawait

# Conflicts:
#	app/ore/permission/PermissionPredicate.scala

Katrix commented on app/controllers/OreBaseController.scala:
```diff
@@ -62,8 +71,14 @@ abstract class OreBaseController(implicit val env: OreEnv,
+
+  def withVersionAsync(versionString: String)(fn: Version => Future[Result])
+                 (implicit request: OreRequest[_], project: Project): Future[Result]
+  = project.versions.find(equalsIgnoreCase[VersionTable](_.versionString, versionString)).flatMap(_.map(fn).getOrElse(Future.successful(notFound())))
```
I'd say you should remove the parenthesis after notFound, it's technically correct as that's what it's defined with, but AFAIK, it does not have any side effects.

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -30,121 +38,194 @@ final class Reviews @Inject()(data: DataHelper,
+    (Authenticated andThen PermissionAction[AuthRequest](ReviewProjects)) async { implicit request =>
+      withProjectAsync(author, slug) { implicit project =>
+        withVersionAsync(versionString) { implicit version =>
+          for {
```
A normal map is probably more clear here

Faithcaio replied:
even better request.user

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -30,121 +38,194 @@ final class Reviews @Inject()(data: DataHelper,
+            case None => Future.successful(NotFound)
+            case Some(review) =>
+              for {
+                _ <- review.setEnded(Timestamp.from(Instant.now()))
```
These can probably be parallelized

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -30,121 +38,194 @@ final class Reviews @Inject()(data: DataHelper,
+            case None => Future.successful(true)
+            case Some(oldreview) =>
+              for {
+                _ <- oldreview.addMessage(Message(this.forms.ReviewDescription.bindFromRequest.get.trim, System.currentTimeMillis(), "takeover"))
```
These can probably be parallelized

Katrix commented on app/controllers/Reviews.scala:
```diff
@@ -30,121 +38,194 @@ final class Reviews @Inject()(data: DataHelper,
+
+          // Then make new one
+          for {
+            _ <- closeOldReview
```
These can probably be parallelized

Katrix commented on app/controllers/Users.scala:
```diff
@@ -68,16 +76,19 @@ class Users @Inject()(fakeUser: FakeUser,
-          val user = this.users.getOrCreate(User.fromSponge(spongeUser)).pullForumData().pullSpongeData()
-          this.redirectBack(request.flash.get("url").getOrElse("/"), user)
+          User.fromSponge(spongeUser).flatMap(this.users.getOrCreate).flatMap { user =>
+            user.pullForumData.flatMap(_.pullSpongeData).flatMap { u =>
```
Add parenthesis back to the mutators here

Katrix commented on app/controllers/Users.scala:
```diff
@@ -121,20 +132,50 @@ class Users @Inject()(fakeUser: FakeUser,
+      case Some(user) =>
+        for {
+          // TODO include orga projects?
+          projectSeq <- service.DB.db.run(queryUserProjects(user).drop(offset).take(pageSize).result)
```
Some of these can probably be parallelized

Katrix commented on app/controllers/Users.scala:
```diff
@@ -171,9 +215,9 @@ class Users @Inject()(fakeUser: FakeUser,
         import writes._
```
Is this still used?

Katrix commented on app/db/impl/access/OrganizationBase.scala:
```diff
@@ -46,53 +50,56 @@ class OrganizationBase(override val service: ModelService,
+        // and should be treated as such.
+        org.toUser.map {
+          case None => throw new IllegalStateException("User not created")
+          case Some(userOrg) => userOrg.pullForumData.flatMap(_.pullSpongeData)
```
Add parenthesis to mutators here.

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -114,45 +125,59 @@ class ProjectBase(override val service: ModelService,
+    } yield {
+      checkArgument(isAvailable, "slug not available", "")
+    }
+    future.flatMap { _ =>
```
Collapse this into a single for comprehension

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -114,45 +125,59 @@ class ProjectBase(override val service: ModelService,
+    project.map { project =>
+      checkArgument(project.id.get == channel.projectId, "invalid project id", "")
+      val checks = for {
+        channels <- project.channels.all
```
These can probably be parallelized

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -114,45 +125,59 @@ class ProjectBase(override val service: ModelService,
+          "last reviewed channel", "")
+      }
+      for {
+        _ <- checks
```
Collapse the previous for comprehension and this one into one

Faithcaio replied:
not sure about that
it currently uses exceptions above to exit
wouldnt it not run the first yield part before deleting the channel?

Katrix replied:
I see no problem.
This:
```scala
val futureRes = for {
  foo1 <- futureFoo1
  foo2 <- futureFoo2
} yield {
  check(foo1.bar)
  foo2
}

for {
  res  <- futureRes
  foo3 <- futureFo3
} yield foo3
```

Has the same effect as this:
```scala
for {
  foo1 <- futureFoo1
  foo2 <- futureFoo2
  res  <- {
    check(foo1.bar)
    Future.successful(foo2)
  }
  foo3 <- futureFoo3
} yield foo3
```

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -161,27 +186,40 @@ class ProjectBase(override val service: ModelService,
 
-    version.remove()
+    val rcUpdate = for {
+      proj <- checks
```
Collapse the previous for comprehension and this one into one

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -161,27 +186,40 @@ class ProjectBase(override val service: ModelService,
+    val rcUpdate = for {
+      proj <- checks
+      rv <- proj.recommendedVersion
+      projects <- proj.versions.sorted(_.createdAt.desc) // TODO optimize: only query one version
```
These can probably be parallelized

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -161,27 +186,40 @@ class ProjectBase(override val service: ModelService,
-    if (channel.versions.isEmpty)
-      this.deleteChannel(channel)
+    val channelCleanup = for {
+      proj <- rcUpdate
```
Collapse the previous for comprehension and this one into one

Katrix commented on app/db/impl/access/ProjectBase.scala:
```diff
@@ -161,27 +186,40 @@ class ProjectBase(override val service: ModelService,
 
-    val versionDir = this.fileManager.getVersionDir(proj.ownerName, project.name, version.name)
-    FileUtils.deleteDirectory(versionDir)
+    channelCleanup.flatMap { proj =>
```
Place this flatMap in the for comprehension

Katrix commented on app/db/impl/access/UserBase.scala:
```diff
@@ -34,9 +41,42 @@ class UserBase(override val service: ModelService,
+    this.withName(name).flatMap {
+      case None => Future.successful(None) // Name not found
+      case Some(toCheck) =>
+        if (user.equals(toCheck)) Future.successful(Some(user)) // Same user
```
Placing the if together with the case here would probably look a bit cleaner

Katrix commented on app/db/impl/access/UserBase.scala:
```diff
@@ -122,13 +156,28 @@ class UserBase(override val service: ModelService,
-  def current(implicit session: Request[_]): Option[User] = session.cookies.get("_oretoken").flatMap { token =>
-    getSession(token.value).map(_.user)
+  def current(implicit session: Request[_], ec: ExecutionContext): Future[Option[User]] = {
+    val asd = session.cookies.get("_oretoken") match {
```
Remove this

Katrix commented on app/models/user/User.scala:
```diff
@@ -283,29 +284,44 @@ case class User(override val id: Option[Int] = None,
-              }
+        this.service.DB.db.run(Project.roleForTrustQuery(pScope.projectId).result).map { l =>
+          val ordering: Ordering[ProjectRole] = Ordering.by(m => m.roleType.trust)
+          l.sorted(ordering).headOption.map(_.roleType.trust).getOrElse(Default)
```
l.sortedBy?

Katrix commented on app/models/user/User.scala:
```diff
@@ -283,29 +284,44 @@ case class User(override val id: Option[Int] = None,
+            val memberTable = TableQuery[OrganizationMembersTable]
+            val roleTable = TableQuery[OrganizationRoleTable]
+
+            // TODO review permission logic
```
Is this TODO resolved?

Faithcaio replied:
maybe?^^

Katrix commented on app/models/user/User.scala:
```diff
@@ -283,29 +284,44 @@ case class User(override val id: Option[Int] = None,
+            }
+
+            service.DB.db.run(query.result).map { l =>
+              l.map(_.trust).find(t => !t.equals(Default)).getOrElse(Default) // Find first non default trust
```
collectFirst

Katrix commented on app/models/user/User.scala:
```diff
@@ -331,10 +347,22 @@ case class User(override val id: Option[Int] = None,
+    this.service.getModelBase(classOf[UserBase]).current.flatMap {
+      case None => Future.successful(false)
+      case Some(user) =>
+        if (user.equals(this)) Future.successful(true)
```
This would probably look nicer if you lifted the comparison into the match.

Katrix commented on app/models/user/User.scala:
```diff
@@ -345,60 +373,58 @@ case class User(override val id: Option[Int] = None,
-      .toSet[RoleType]
-    this
+      .toSet[RoleType])
+    Future.successful(this)
```
Any reason for returning a Future here?

Faithcaio replied:
DB updates are done above ; should get future from that

Katrix commented on app/models/user/User.scala:
```diff
@@ -345,60 +373,58 @@ case class User(override val id: Option[Int] = None,
-    }.foreach(this.avatarUrl = _)
-    this
+    }.foreach(this.setAvatarUrl(_))
+    Future.successful(this)
```
Any reason for returning a Future here?

Katrix commented on app/models/user/User.scala:
```diff
@@ -532,14 +561,22 @@ case class User(override val id: Option[Int] = None,
-    hasFlags || hasReview || hasNotifications
+
+    for {
+      hasFlags <- this can ReviewFlags in GlobalScope
```
These can probably be parallelized

Katrix commented on app/models/project/ProjectSettings.scala:
```diff
@@ -138,59 +144,83 @@ case class ProjectSettings(override val id: Option[Int] = None,
-        move(pendingPath, iconDir.resolve(pendingPath.getFileName))
+    val ownerSet = formData.ownerId.find(_ != project.ownerId) match {
+      case None => Future.successful(true)
+      case Some(ownerId) => this.userBase.get(ownerId).flatMap(user => project.setOwner(user.get))}
```
Place the curly brace on a new line

Katrix commented on app/models/project/Version.scala:
```diff
@@ -253,9 +250,17 @@ case class Version(override val id: Option[Int] = None,
+    if (this.projectId == -1) Future.successful(false)
+    else {
+      for {
+        hashExists <- this.schema.hashExists(this.projectId, this.hash)
```
These can be parallelized

Katrix commented on app/models/user/Organization.scala:
```diff
@@ -9,12 +9,17 @@ import db.impl.model.OreModel
 import ore.user.{MembershipDossier, UserOwned}
 import ore.{Joinable, Visitable}
+import slick.lifted.{Compiled, Rep, TableQuery}
+import db.impl.OrePostgresDriver.api._
```
Duplicate import

Katrix commented on app/models/user/Organization.scala:
```diff
@@ -55,9 +60,22 @@ case class Organization(override val id: Option[Int] = None,
+    override def getTrust(user: User)(implicit ex: ExecutionContext): Future[Trust] = {
+      this.userBase.service.DB.db.run(Organization.roleForTrustQuery(id.get).result).map { l =>
+        val ordering: Ordering[OrganizationRole] = Ordering.by(m => m.roleType.trust)
+        l.sorted(ordering).headOption.map(_.roleType.trust).getOrElse(Default)
```
l.sortedBy?

Katrix commented on app/models/viewhelper/VersionData.scala:
```diff
@@ -0,0 +1,44 @@
+
+
+  def filteredDependencies = {
+    dependencies.filterNot(_._1.pluginId.equals(SpongeApiId))
```
Use == instead of equals for these

Katrix commented on app/models/viewhelper/UserData.scala:
```diff
@@ -0,0 +1,73 @@
+
+  def pgpFormCall = {
+    user.pgpPubKey.map { _ =>
+      routes.Users.verify(Some(routes.Users.deletePgpPublicKey(user.name, None, None).path()))
```
Remove parenthesis from path here

Katrix commented on app/models/viewhelper/UserData.scala:
```diff
@@ -0,0 +1,73 @@
+
+  def of[A](request: OreRequest[A], user: User)(implicit cache: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext, service: ModelService): Future[UserData] = {
+    for {
+      isOrga <- user.isOrganization
```
These can be parallelized

Katrix commented on app/models/viewhelper/ScopedProjectData.scala:
```diff
@@ -0,0 +1,54 @@
+  def of(currentUser: Option[User], project: Project)(implicit ec: ExecutionContext, cache: AsyncCacheApi): Future[ScopedProjectData] = {
+    currentUser.map { user =>
+      for {
+        projectOwner <- project.owner.user
```
These can be parallelized

Katrix commented on app/models/viewhelper/ScopedOrganizationData.scala:
```diff
@@ -0,0 +1,40 @@
+    }
+  }
+
+  def of[A](currentUser: Option[User], orga: Option[Organization])(implicit cache: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext, service: ModelService): Future[Option[ScopedOrganizationData]] = {
```
Newline somewhere in here

Katrix commented on app/models/viewhelper/ScopedOrganizationData.scala:
```diff
@@ -0,0 +1,40 @@
+
+  def cacheKey(orga: Organization, user: User) = s"""organization${orga.id.get}foruser${user.id.get}"""
+
+  def of[A](currentUser: Option[User], orga: Organization)(implicit cache: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext, service: ModelService): Future[ScopedOrganizationData] = {
```
Newline somewhere in here

Katrix commented on app/models/viewhelper/ProjectData.scala:
```diff
@@ -0,0 +1,136 @@
+                       noteCount: Int, // getNotes.size
+                       lastVisibilityChange: Option[VisibilityChange],
+                       lastVisibilityChangeUser: String // users.get(project.lastVisibilityChange.get.createdBy.get).map(_.username).getOrElse("Unknown")
+                      ) extends JoinableData[ProjectRole, ProjectMember, Project]()
```
Remove parenthesis on extended class

Katrix commented on app/models/viewhelper/ProjectData.scala:
```diff
@@ -0,0 +1,136 @@
+                       lastVisibilityChange: Option[VisibilityChange],
+                       lastVisibilityChangeUser: String // users.get(project.lastVisibilityChange.get.createdBy.get).map(_.username).getOrElse("Unknown")
+                      ) extends JoinableData[ProjectRole, ProjectMember, Project]()
+{
```
Place this on the previous line

Katrix commented on app/models/viewhelper/ProjectData.scala:
```diff
@@ -0,0 +1,136 @@
+    val projectOwner = request.data.currentUser.get
+
+    val settings = project.settings
+    val ownerRole = null // TODO?
```
Is this TODO resolved?

Faithcaio replied:
This is the default data for pending projects.
This value is probably not accessed.

Katrix commented on app/models/viewhelper/ProjectData.scala:
```diff
@@ -0,0 +1,136 @@
+
+    implicit val userBase = project.userBase
+    for {
+      settings <- project.settings
```
These can be parallelized

Katrix commented on app/models/viewhelper/OrganizationData.scala:
```diff
@@ -0,0 +1,49 @@
+  def of[A](orga: Organization)(implicit cache: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext, service: ModelService): Future[OrganizationData] = {
+    implicit val users = orga.userBase
+    for {
+      role <- orga.owner.headRole
```
These can be parallelized

Katrix commented on app/models/viewhelper/OrganizationData.scala:
```diff
@@ -0,0 +1,49 @@
+
+  def cacheKey(orga: Organization) = "organization" + orga.id.get
+
+  def of[A](orga: Organization)(implicit cache: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext, service: ModelService): Future[OrganizationData] = {
```
Newlines somewhere in here

Katrix commented on app/models/viewhelper/OrganizationData.scala:
```diff
@@ -0,0 +1,49 @@
+  }
+
+
+  def of[A](orga: Option[Organization])(implicit cache: AsyncCacheApi, db: JdbcBackend#DatabaseDef, ec: ExecutionContext, service: ModelService): Future[Option[OrganizationData]] = {
```
Newlines somewhere in here

Katrix commented on app/models/viewhelper/HeaderData.scala:
```diff
@@ -0,0 +1,163 @@
+
+  def hasUser = currentUser.isDefined
+
+  def isCurrentUser(userId: Int) = currentUser.map(_.id).contains(userId)
```
Comparing unrelated types. Probably meant to use flatMap instead of map

Katrix commented on app/models/viewhelper/HeaderData.scala:
```diff
@@ -0,0 +1,163 @@
+                  HideProjects -> false,
+                  HardRemoveProject -> false,
+                  UserAdmin -> false,
+    HideProjects -> false)
```
Not aligned with the others

Katrix commented on app/models/viewhelper/HeaderData.scala:
```diff
@@ -0,0 +1,163 @@
+  private def getHeaderData(user: User)(implicit ec: ExecutionContext, db: JdbcBackend#DatabaseDef) = {
+
+    for {
+      perms <- perms(Some(user))
```
These can be parallelized

Katrix commented on app/models/viewhelper/HeaderData.scala:
```diff
@@ -0,0 +1,163 @@
+    else {
+      val user = currentUser.get
+      for {
+        reviewFlags       <- user can ReviewFlags in GlobalScope map ((ReviewFlags, _))
```
These can be parallelized

Katrix commented on app/ore/organization/OrganizationMember.scala:
```diff
@@ -14,15 +16,22 @@ import ore.user.Member
+    * @return Top role
+    */
+  override def headRole(implicit ec: ExecutionContext): Future[OrganizationRole] =
+    this.roles.map(role => role.max(ordering))
```
role.maxBy?

Katrix commented on app/ore/permission/PermissionPredicate.scala:
```diff
@@ -5,34 +5,59 @@ import models.project.Project
+            orgTest <- org.scope.test(user, p)
+            projectTest <- project.scope.test(user, p)
+          } yield {
+            orgTest | projectTest
```
Does this need to be a bitwise or operation? If not then you only need one of the futures.

Katrix commented on app/ore/project/ProjectMember.scala:
```diff
@@ -16,22 +18,27 @@ import ore.user.Member
+    *
+    * @return Top role
+    */
+  override def headRole(implicit ec: ExecutionContext): Future[ProjectRole] = this.roles.map(_.max(ordering))
```
maxBy?

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -109,16 +106,27 @@ trait ProjectFactory {
-    Right(version)
+      return Future.successful(Left("error.version.invalidPluginId"))
+    val version = for {
+      channels <- project.channels.all
```
These can be parallelized

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -243,36 +252,47 @@ trait ProjectFactory {
-        message = messages("notification.project.invite", role.roleType.title, project.name)
-      ))
+    val checks = for {
+      exists <- this.projects.exists(project)
```
These can be parallelized

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -283,14 +303,21 @@ trait ProjectFactory {
+      checkState(channelCount < this.config.projects.get[Int]("max-channels"), "channel limit reached", "")
+    }
+
+    checks.flatMap { _ =>
```
Do this in the for comprehension

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -299,82 +326,105 @@ trait ProjectFactory {
+
+    val channel = for {
+      // Create channel if not exists
+      channel <- getOrCreateChannel(pending, project)
```
These can be parallelized

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -299,82 +326,105 @@ trait ProjectFactory {
-          newVersion.addTag(tag)
-        }
-      }
+    val newVersion = channel.flatMap { channel =>
```
Do this in the for comprehension

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -299,82 +326,105 @@ trait ProjectFactory {
-    addTags(ForgeId, "Forge", TagColors.Forge)
+    for {
+      channel <- channel
+      newVersion <- newVersion
```
Collapse this for comprehension and the previous one

Katrix commented on app/ore/project/factory/ProjectFactory.scala:
```diff
@@ -299,82 +326,105 @@ trait ProjectFactory {
+    for {
+      channel <- channel
+      newVersion <- newVersion
+      spongeTag <- addTags(newVersion, SpongeApiId, "Sponge", TagColors.Sponge)
```
This can be parallelized

Katrix commented on app/ore/rest/OreRestfulApi.scala:
```diff
@@ -41,18 +48,131 @@ trait OreRestfulApi {
+      }
+      //categories.map(_.toSeq).map { cats =>
+      val filtered = cats.map { ca =>
+        query.filter { case (p, v, c) =>
```
Mark v and c as ignored

Katrix commented on app/security/spauth/SpongeAuthApi.scala:
```diff
@@ -9,7 +9,7 @@ import play.api.libs.functional.syntax._
 import util.WSUtils.parseJson
 import play.api.libs.json._
-import play.api.libs.ws.{WSClient, WSResponse}
+import play.api.libs.ws.{WSClient, WSRequest, WSResponse}
```
Unused import. Also remove Await further down.

Katrix reviewed (changes requested):
Do a favor and run optimize imports on changed files? I counted a lot of unused imports that you've added when I went through everything.

For the stuff that can be parallelized. If you want I can add them together with EitherT and OptionT instead.

I should also say that I haven't looked at the views in any way. Someone else should probably do that.

Faithcaio committed:
adressing some issues

Faithcaio committed:
auto import cleanup

Faithcaio committed:
fix compile errors

felixoi reviewed (commented):
- [x] If you change the project visibility in your branch the visiblity is changed but the view is not updated. I've to reload, I think in master the view changes automatically
- [x] Remove filter button is always visible also if no categories are selected [![https://gyazo.com/7d0390cfa4f4de055fca0be7954ff1ed](https://i.gyazo.com/7d0390cfa4f4de055fca0be7954ff1ed.png)](https://gyazo.com/7d0390cfa4f4de055fca0be7954ff1ed)
- [x] Adding members to a project is not working (always outputs user not found)
- [x] Adding comments to a review is not working
- [x] Flag history is broken. It displays there is one entry in the dropdown but opening the history returns a blank one.
- [x] Child pages are wrongly displayed in the side nav
- [x] Slugs of child pages are wrong

Faithcaio committed:
fix pages
fix category filter clearing always being visible
and some other stuff not getting displayed

Faithcaio committed:
fix REST users

Faithcaio committed:
wait for visibility change before returning

Faithcaio committed:
fix page links in REST

Faithcaio commented:
project avatar seems to be broken on ore live too

progwml6 commented:
file issues please for future work that needs to be done

felixoi reviewed (dismissed):
I think I've tested mostly everything and I couldn't find more bugs.

Faithcaio committed:
fix page saving

Faithcaio committed:
fix organization owner

Faithcaio committed:
maybe fix project permission check

Faithcaio committed:
maybe fix project permission check more

Faithcaio committed:
fix user organizations not showing up

Faithcaio committed:
fix membersList when logged out

Faithcaio committed:
fix user permissions
